### PR TITLE
Add an improved "forward extremities" metric

### PR DESCRIPTION
**Pull request by richvdh**
Hopefully, `(N(extremities)-1) * N(state_events)` is a more realistic approximation to "how big a problem is this room?".

In order to get there, I've replaced `BucketCollector` with something else. I think it's a bit easier to use.

### CODE REVIEW

File: synapse/storage/databases/main/metrics.py [hunk 1]
```diff
@@ -12,17 +12,35 @@
...
 from synapse.storage.databases.main.event_push_actions import (
     EventPushActionsWorkerStore,
 )
 
+# Collect metrics on the number of forward extremities that exist.
+_extremities_collecter = GaugeBucketCollector(
+    "synapse_forward_extremities",
+    "Number of rooms on the server with the given number of forward extremities"
+    " or fewer",
+    buckets=[1, 2, 3, 5, 7, 10, 15, 20, 50, 100, 200, 500],
```
Comment (ara4n) [inline synapse/storage/databases/main/metrics.py:29]: why would we ever expect there to be significantly more than 10 forward extremities (or whatever the dummy threshold is)?

Comment (richvdh) [reply synapse/storage/databases/main/metrics.py:29]: well, this is what we have today...

There are a few reasons we might end up with more:
 * there are no users in the room who can send an event, for reasons of power-levels or consent
 * we can't keep up with the amount of divergence in the dag (#7792 was a big cause of this, but is hopefully now fixed)

File: synapse/metrics/__init__.py [hunk 2]
```diff
@@ -205,63 +206,81 @@ def _register_with_collector(self):
...
 
-    Args:
-        name (str): Base name of metric to be exported to Prometheus.
-        data_collector (callable -> dict): A synchronous callable that
-            returns a dict mapping bucket to number of items in the
-            bucket. If these buckets are not the same as the buckets
-            given to this class, they will be remapped into them.
-        buckets (list[float]): List of floats/ints of the buckets to
-            give to Prometheus. +Inf is ignored, if given.
+    The data is updated by calling `update_data` with an itrerable of measurements.
```
Comment (anoadragon453) [inline synapse/metrics/__init__.py:212]:
```suggestion
    The data is updated by calling `update_data` with an iterable of measurements.
```

File: synapse/metrics/__init__.py [hunk 3]
```diff
@@ -205,63 +206,81 @@ def _register_with_collector(self):
...
+                if v <= bound:
+                    bucket_values[i] += 1
+                    break
+
+            # ... and increment the sum
+            total += v
+
+        # now, aggregate the bucket values so that they count the number of entries in
+        # that bucket or below.
+        bucket_values = itertools.accumulate(bucket_values)
```
Comment (anoadragon453) [inline synapse/metrics/__init__.py]: This may be standard practice... but why do we do this rather than just having the individual count per bucket?

Comment (richvdh) [reply synapse/metrics/__init__.py]: because that's how prometheus histograms work, basically...

**Review by erikjohnston — APPROVED** [review]

File: synapse/metrics/__init__.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/metrics/__init__.py]: Fair enough!

**Comment by richvdh** [comment]
I'm going to rebase this into atomic commits that I can then merge into develop.

**Pull request merged**